packages/common-cosmos/src/codegen/codegen-controller.ts

```diff
@@ -0,0 +1,120 @@
+// Copyright 2020-2023 SubQuery Pte Ltd authors & contributors
+// SPDX-License-Identifier: GPL-3.0
+
+import fs from 'fs';
+import path from 'path';
+import telescope from '@cosmology/telescope';
+import {makeTempDir} from '@subql/common';
+import {Data} from 'ejs';
+import {copySync} from 'fs-extra';
+import {CosmosChainType} from '../project';
+import {TELESCOPE_OPTS} from './constants';
+
+const PROTO_INTERFACES_ROOT_DIR = 'src/types/proto-interfaces';
+const PROTO_INTERFACE_TEMPLATE_PATH = path.resolve(__dirname, './templates/proto-interface.ts.ejs');
+const TYPE_ROOT_DIR = 'src/types';
+
+interface ProtobufRenderProps {
+  messageNames: string[]; // all messages
+  path: string; // should process the file Path and concat with PROTO dir
+}
+
+export function processProtoFilePath(path: string): string {
+  // removes `./proto` and `.proto` suffix, converts all `.` to `/`
+  // should be able to accept more paths, not just from `proto directory`
+  return `./proto-interfaces/${path.replace(/^\.\/proto\/|\.proto$/g, '').replace(/\./g, '/')}`;
+}
+
+export function isProtoPath(filePath: string, projectPath: string): boolean {
+  // check if the protobuf files are under ./proto directory
+  return !!path.join(projectPath, filePath).startsWith(path.join(projectPath, './proto/'));
+}
+
+export function prepareProtobufRenderProps(
+  chainTypes: Map<string, CosmosChainType>[],
+  projectPath: string
+): ProtobufRenderProps[] {
+  return chainTypes.flatMap((chainType) => {
+    return Object.entries(chainType).map(([key, value]) => {
+      const filePath = path.join(projectPath, value.file);
+      if (!fs.existsSync(filePath)) {
+        throw new Error(`Error: chainType ${key}, file ${value.file} does not exist`);
+      }
+      if (!isProtoPath(value.file, projectPath)) {
+        console.error(
+          `Codegen will not apply for this file: ${value.file} Please ensure it is under the ./proto directory`
+        );
+      }
+      return {
+        messageNames: value.messages,
+        path: processProtoFilePath(value.file),
+      };
+    });
+  });
+}
+
+export async function tempProtoDir(projectPath: string): Promise<string> {
+  const tmpDir = await makeTempDir();
+  const userProto = path.join(projectPath, './proto');
+  const commonProtoPaths = [
+    require('@protobufs/amino'),
+    require('@protobufs/confio'),
+    require('@protobufs/cosmos'),
+    require('@protobufs/cosmos_proto'),
+    require('@protobufs/gogoproto'),
+    require('@protobufs/google'),
+    require('@protobufs/ibc'),
+    require('@protobufs/tendermint'),
+  ];
```

Comment:
Doesn't this load the files? Dont we want `require.resolve`?

Reply:
The packages exports its path
https://github.com/cosmology-tech/proto-registry/blob/main/packages/confio/index.js